### PR TITLE
`Slider.onChangeStart` & `Slider.onChangeEnd` are not called on keyboard shortcuts

### LINKED ISSUE #123315: Slider `onChangeEnd` is never called when arrow keys are used.
## Steps to Reproduce

1. Execute `flutter run` on the code sample 
2. Click on the slider so it gains focus
3. Use left and right arrow keys to change slider position
4. Note that `onChangeEnd` is never called.  Only tested on Windows 10.

**Expected results:** `onChangeEnd` should be called at the end of every slider change, regardless how it is triggered.  With arrow keys, I would expect `onChangeEnd` to be called when the arrow key is released.

**Actual results:** `onChangeEnd` is never called when the arrow keys are used.

<details>
<summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(const SliderApp());

class SliderApp extends StatelessWidget {
  const SliderApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: SliderExample(),
    );
  }
}

class SliderExample extends StatefulWidget {
  const SliderExample({super.key});

  @override
  State<SliderExample> createState() => _SliderExampleState();
}

class _SliderExampleState extends State<SliderExample> {
  double _currentSliderValue = 20;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Slider')),
      body: Slider(
        value: _currentSliderValue,
        max: 100,
        divisions: 5,
        label: _currentSliderValue.round().toString(),
        onChanged: (double value) {
          setState(() {
            _currentSliderValue = value;
          });
        },
        onChangeEnd: (double value) {
          print('onChangeEnd');
        },
      ),
    );
  }
}
```

</details>

### PR DESCRIPTION
**Pull request by dleyba042**
fixes https://github.com/flutter/flutter/issues/123315
--------

This PR makes changes to the _actionHandler function used on the Slider.Dart Widget for Key Events. It ensures onChangeStart is called at the beginning of a Key Event and onChangeEnd at the end of one. This PR includes a test for the changes made.
I ran all existing tests after my changes were made and they passed.

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by TahaTesser** [comment]
Hi @dleyba042 
Please note: https://github.com/flutter/flutter/issues/123315#issuecomment-1532966760

Thanks for attempting to fix this, however several issues in your implementation. 

Check out the [Style guide for Flutter repo](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo) and see how `increaseAction` and `decreaseAction` is currently calling `onChanged`.

in order to fix this, first we need to move the `clampDouble`s  from `onChanged` to separate functions then reuse it for the callback including in the action handler. 

For tests, please check out other similar tests in the framework on how to title, format and test the functionality.

**Comment by TahaTesser** [comment]
Hi @dleyba042 
Do you've plans to update this PR? :)

**Comment by dleyba042** [comment]
Hi @TahaTesser, yes, I made some changes based on your comments. I'm not sure at this time how to fix the Linux errors that are happening. If you wanted to use any part of what I have done in your fix, be my guest.

**Comment by TahaTesser** [comment]
> Hi @dleyba042 Please note: [#123315 (comment)](https://github.com/flutter/flutter/issues/123315#issuecomment-1532966760)
> 
> Thanks for attempting to fix this, however several issues in your implementation.
> 
> Check out the [Style guide for Flutter repo](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo) and see how `increaseAction` and `decreaseAction` is currently calling `onChanged`.
> 
> in order to fix this, first we need to move the `clampDouble`s from `onChanged` to separate functions then reuse it for the callback including in the action handler.
> 
> For tests, please check out other similar tests in the framework on how to title, format and test the functionality.

As mentioned, these tests had several issues including some redundancy. Nevertheless given it is your first PR. I've fixed the tests for you.

**Review by TahaTesser — CHANGES_REQUESTED** [review]
I just pushed rewritten tests. Jus this extra space needs to be removed.

**Review by TahaTesser — CHANGES_REQUESTED** [review]
I just pushed rewritten tests. For this change, we don't need a new set of tests, the existing keyboard shortcuts are already testing `onChanged` callback. I just update these tests.

Finally, just this extra space needs to be removed.

**Review by TahaTesser — CHANGES_REQUESTED** [review]

**Review by TahaTesser — APPROVED** [review]
LGTM

**Review by TahaTesser — APPROVED** [review]
LGTM

**Comment by TahaTesser** [comment]
@Piinks 
This needs a second review

**Review by Piinks — APPROVED** [review]
Hey @dleyba042, welcome! Thank you for contributing!
Appreciate the review @TahaTesser. This also LGTM! 🎉

**Comment by dleyba042** [comment]
@TahaTesser sorry for the slow response currently on vacation. Thank you for the input and the learning opportunity! Thanks @Piinks it was a great learning experience.

**Pull request merged**
